lib/metadata/chordMetadata.hpp

```diff
@@ -392,6 +399,128 @@ class chordMetadata : public metadataObject {
+    }
+
+    /// The number of bad inputs in the RFI systems bad input list.
+    /// This value is mostly needed for renormalization of the SK values.
+    uint32_t get_rfi_num_bad_inputs() const {
+        return metadata[jsonMetadata::RFI_NUM_BAD_INPUTS].template get<uint32_t>();
+    }
+
+    void set_rfi_num_bad_inputs(const uint32_t rfi_num_bad_inputs) {
+        metadata[jsonMetadata::RFI_NUM_BAD_INPUTS] = rfi_num_bad_inputs;
+    }
+
+    /// The number of FPGA frames flagged as containing RFI.
+    /// NOTE: This value might contain overlap with lost samples, so it can count
+    /// missing samples as samples with RFI.  For renormalization this value
+    /// should NOT be used, use @c lost_timesamples instead.
+    /// This value will be filled even if RFI zeroing is disabled.
+    int32_t get_rfi_flagged_samples() const {
+        return metadata[jsonMetadata::RFI_FLAGGED_SAMPLES].template get<int32_t>();
+    }
+
+    void set_rfi_flagged_samples(const int32_t flagged_samples)  {
+        // very much non-atomic, due to json dict entry creation
+        metadata[jsonMetadata::RFI_FLAGGED_SAMPLES] = flagged_samples;
+    }
+
+    int32_t get_lost_timesamples() const {
+        return metadata[jsonMetadata::LOST_TIMESAMPLES].template get<int32_t>();
+    }
+
+    void set_lost_timesamples(int32_t lost_timesamples) {
+        // very much non-atomic, due to json dict entry creation
+        metadata[jsonMetadata::LOST_TIMESAMPLES] = lost_timesamples;
+    }
+
+    void atomic_add_lost_timesamples(const int32_t lost_samples) {
+        // RH: this is almost certainly not admissible code, but also almost certain, "works".
+        // RH: this is not actually atomic and has race conditions if the underlying json dict changes
+        static_assert(std::is_same<std::int64_t, nlohmann::json::number_integer_t>::value, "Roland's horrible hack fails");
+        *reinterpret_cast<std::atomic_int64_t*>(metadata[jsonMetadata::LOST_TIMESAMPLES].template get_ptr<std::int64_t*>()) += lost_samples;
+    }
+
+    // non-science metadata
+
+   timeval get_first_packet_recv_time() const {
+        return metadata[jsonMetadata::FIRST_PACKET_RECV_TIME].template get<timeval>();
+   }
+
+   void set_first_packet_recv_time(const timeval time_v) {
+        metadata[jsonMetadata::FIRST_PACKET_RECV_TIME] = time_v;
+   }
+
+    // links to other data
+
+    stream_t get_stream_id() const {
+        return stream_t{.id = metadata[jsonMetadata::STREAM_ID].template get<uint64_t>()};
+    }
+
+    void set_stream_id(const stream_t stream_id) {
+        metadata[jsonMetadata::STREAM_ID] = stream_id.id;
+    }
+
+    /// ID of the dataset
+    dset_id_t get_dataset_id() const {
+        return metadata[jsonMetadata::DATASET_ID].template get<dset_id_t>();
+    }
+
+    void set_dataset_id(const dset_id_t dset_id) {
+        metadata[jsonMetadata::DATASET_ID] = dset_id;
+    }
+
+private:
+    jsonMetadata::metadata metadata;
```

Comment:
We will need serialization/deserialization functions for this class.   Specifically for stages like `bufferSend`/`bufferRecv` we currently only perform "shallow copies" which worked for the flat struct model, but don't work with this dynamic structure.   Of course for CHIME compatibility we'll also want to have a "legacy" version, which will likely have some conflicts with what Jim is working on now...